Mathlib/AlgebraicGeometry/Morphisms/Constructors.lean

```diff
@@ -56,136 +56,157 @@ instance AffineTargetMorphismProperty.diagonal_respectsIso (P : AffineTargetMorp
-  · rintro ⟨𝒰, h𝒰, H⟩
-    refine ⟨𝒰, inferInstance, fun _ => Scheme.affineCover _, inferInstance, ?_⟩
-    intro i j k
-    apply H
-  tfae_have 5 → 1
-  · rintro ⟨𝒰, _, 𝒰', _, H⟩
-    exact diagonal_targetAffineLocally_of_openCover P hP f 𝒰 𝒰' H
-  tfae_finish
-#align algebraic_geometry.affine_target_morphism_property.is_local.diagonal_affine_open_cover_tfae AlgebraicGeometry.AffineTargetMorphismProperty.IsLocal.diagonal_affine_openCover_TFAE
-
-theorem AffineTargetMorphismProperty.IsLocal.diagonal {P : AffineTargetMorphismProperty}
-    (hP : P.IsLocal) : P.diagonal.IsLocal :=
-  have := hP.1
-  AffineTargetMorphismProperty.isLocalOfOpenCoverImply P.diagonal
-    fun {_ _} f => ((hP.diagonal_affine_openCover_TFAE f).out 1 3).mp
-#align algebraic_geometry.affine_target_morphism_property.is_local.diagonal AlgebraicGeometry.AffineTargetMorphismProperty.IsLocal.diagonal
-
-theorem diagonal_targetAffineLocally_eq_targetAffineLocally (P : AffineTargetMorphismProperty)
-    (hP : P.IsLocal) : (targetAffineLocally P).diagonal = targetAffineLocally P.diagonal := by
-  ext _ _ f
-  exact ((hP.diagonal_affine_openCover_TFAE f).out 0 1).trans
-    ((hP.diagonal.affine_openCover_TFAE f).out 1 0)
-#align algebraic_geometry.diagonal_target_affine_locally_eq_target_affine_locally AlgebraicGeometry.diagonal_targetAffineLocally_eq_targetAffineLocally
+  rw [← Q.cancel_left_of_respectsIso (pullbackDiagonalMapIso f _ f₁ f₂).hom]
+  convert HasAffineProperty.of_isPullback (P := P) (.of_hasPullback _ _) H
+  · apply pullback.hom_ext <;> simp
+  · infer_instance
+  · infer_instance
+#align algebraic_geometry.affine_target_morphism_property.diagonal_of_target_affine_locally AlgebraicGeometry.HasAffineProperty.affineProperty_diagonal_of_diagonal_of_isPullback
+
+lemma _root_.CategoryTheory.IsPullback.of_id_fst {C} [Category C] {X Y : C} (f : X ⟶ Y) :
+    IsPullback (𝟙 _) f f (𝟙 _) := IsPullback.of_horiz_isIso ⟨by simp⟩
+
+lemma _root_.CategoryTheory.IsPullback.of_id_snd {C} [Category C] {X Y : C} (f : X ⟶ Y) :
+    IsPullback f (𝟙 _) (𝟙 _) f := IsPullback.of_vert_isIso ⟨by simp⟩
+
+theorem HasAffineProperty.of_affineProperty_diagonal
+    (P) {Q} [HasAffineProperty P Q] {X Y} {f : X ⟶ Y} [IsAffine Y] :
+    Q.diagonal f ↔ P.diagonal f := by
+  letI := isLocal_affineProperty P
+  refine ⟨fun hf ↦ ?_, affineProperty_diagonal_of_diagonal_of_isPullback P (.of_id_fst _)⟩
+  rw [← Q.diagonal.cancel_left_of_respectsIso
+    (pullback.fst (f := f) (g := 𝟙 Y)), pullback.condition, Category.comp_id] at hf
+  let 𝒰 := X.affineCover.pushforwardIso (inv (pullback.fst (f := f) (g := 𝟙 Y)))
+  have (i) : IsAffine (𝒰.obj i) := by dsimp [𝒰]; infer_instance
+  exact HasAffineProperty.diagonal_of_openCover P f (Scheme.openCoverOfIsIso (𝟙 _))
+    (fun _ ↦ 𝒰) (fun _ _ _ ↦ hf _ _)
+
+theorem HasAffineProperty.affineProperty_diagonal (P) {Q} [HasAffineProperty P Q] :
+    Q.diagonal = AffineTargetMorphismProperty.of P.diagonal := by
+  ext X Y f _
+  exact of_affineProperty_diagonal P
```

Comment:
I think this is not needed as it follows from the `HasAffineProperty` instance below and the general `HasAffineProperty.iff` (if you want you can state the `[HasAffineProperty P Q] : Q = AffineTargetMorphismProperty.of P` separately from the `HasAffineProperty.iff` lemma).